substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVLineRecord.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2020, 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, 2020, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.objectfile.pecoff.cv;
+
+import com.oracle.objectfile.ObjectFile;
+
+import java.util.ArrayList;
+
+/*
+ * A line record (DEBUG_S_LINES) consists of a list of (file block record + subrecords).
+ * Graal will generate one CVLineRecord per function.
+ */
+final class CVLineRecord extends CVSymbolRecord {
+
+    /* Header: addr (4 bytes):section (2 bytes) flags (2 bytes) chunck length (4 bytes). */
+    private static final int LINE_RECORD_HEADER_SIZE = Integer.BYTES + Short.BYTES * 2 + Integer.BYTES;
+
+    private static final int DEFAULT_LINE_BLOCK_COUNT = 100;
+    private static final int DEFAULT_LINE_ENTRY_COUNT = 100;
+
+    /* Has columns flag = 0x80 - not supported. */
+    private static final short CB_HAS_NO_COLUMNS_FLAG = 0x00;
+
+    private String symbolName;
+    private ArrayList<FileBlock> fileBlocks = new ArrayList<>(DEFAULT_LINE_BLOCK_COUNT);
+
+    CVLineRecord(CVDebugInfo cvDebugInfo, String symbolName) {
+        super(cvDebugInfo, CVDebugConstants.DEBUG_S_LINES);
+        this.symbolName = symbolName;
+    }
+
+    void addNewFile(int fileId) {
+        fileBlocks.add(new FileBlock(fileId));
+    }
+
+    void addNewLine(int addr, int line) {
+        fileBlocks.get(fileBlocks.size() - 1).addEntry(new LineEntry(addr, line));
+    }
+
+    int getCurrentFileId() {
+        return fileBlocks.isEmpty() ? 0 : fileBlocks.get(fileBlocks.size() - 1).fileId;
```

Comment:
Since the builder already ensures that this is only called if `fileBlocks` is not empty, this is not a problem at the moment, but I think that `0` may not be the best choice because it is a valid `fileId`.

Perhaps it would be better to simply assert here that `fileBlocks` is not empty and expect the caller to take care of it.
```suggestion
        assert !fileBlocks.isEmpty();
        return fileBlocks.get(fileBlocks.size() - 1).fileId;
```

Reply:
Done.